osquery/database/plugins/rocksdb.cpp

```diff
@@ -114,6 +120,41 @@ class RocksDBDatabasePlugin : public DatabasePlugin {
 /// Backing-storage provider for osquery internal/core.
 REGISTER_INTERNAL(RocksDBDatabasePlugin, "database", "rocksdb");
 
+class RocksDBGuard {
+ public:
+  RocksDBGuard() : my_tid_(platformGetTid()), locked_(false) {}
+
+  bool safeEnter() {
+    if (!owner_tid_is_valid_ || owner_tid_ != my_tid_) {
+      mu_.lock();
```

Comment:
So, at root is the statement that "The API allows for concurrent reads, writes, deletes, etc, all handled by RocksDB." There is an assumption that underlies this statement - that a RocksDB API itself will never turn around and call another RocksDB API. The logging callback, unfortunately, violates this assumption.

So, to carry your example, we have to handle the following situation:

Rocks DB Background Thread -> Flush -> Rocks DB memtable lock -> Log Callback -> put -> attempt put lock
Thread X -> Put -> put lock -> attempt memtable lock

There is an additional, higher-level problem as well. The challenge is that we have a second multi-threaded "black-box", Google Logging, and we have no way to control the locking behavior of either black box.

So, consider:

Thread x->Rocks DB->Rocks DB mutex->Logging-callback->Google Logging->Google mutex
Thread y->Google Logging->Google mutex->Rocks DB->Rock DB mutex

So, in my testing, it turned out to be insufficient to prevent single-thread re-entrancy. In my first cut at this solution, I simply had a per-thread guard (this is relatively easy to do - basically, make the static members of RocksDBGuard non-static, and you end up with a per-thread guard). I ran into both the above deadlocks with that version.

The alternate solution I explored was to essentially redesign the BufferedLogSink class, so that it's send method always buffers the log and never directly calls the logger, and an independent thread regularly flushes the buffer. This will clearly prevent the deadlock between Google Logging and RocksDB by making sure that Google Logging's callback always terminates with a buffer operation. As I considered this option, though, it seemed to me I would end up re-creating a chunk of what RocksDB does in the first place.

Perhaps a better way to go is to make the fix in RocksDB - perhaps it can buffer it's logging, and only call the logging callback from a background thread - as far as I can see, it has most of the mechanisms to pull this off.